src/util/bip143.rs

```diff
@@ -99,6 +99,117 @@ impl SighashComponents {
     }
 }
 
+/// A replacement for SigHashComponents which supports all sighash modes
+pub struct SigHashCache<'a> {
+    /// Access to transaction required for various introspection
+    tx: &'a Transaction,
+    /// Hash of all the previous outputs, computed as required
+    hash_prevouts: Option<sha256d::Hash>,
+    /// Hash of all the input sequence nos, computed as required
+    hash_sequence: Option<sha256d::Hash>,
+    /// Hash of all the outputs in this transaction, computed as required
+    hash_outputs: Option<sha256d::Hash>,
+}
+
+impl<'a> SigHashCache<'a> {
+    /// Compute the sighash components from an unsigned transaction and auxiliary
+    /// in a lazy manner when required.
+    /// For the generated sighashes to be valid, no fields in the transaction may change except for
+    /// script_sig and witnesses.
+    pub fn new(tx: &Transaction) -> SigHashCache {
+        SigHashCache {
+            tx: tx,
+            hash_prevouts: None,
+            hash_sequence: None,
+            hash_outputs: None,
+        }
+    }
+
+    /// Calculate hash for prevouts
+    pub fn hash_prevouts(&mut self) -> sha256d::Hash {
+        if self.hash_prevouts.is_none() {
+            let mut enc = sha256d::Hash::engine();
+            for txin in &self.tx.input {
+                txin.previous_output.consensus_encode(&mut enc).unwrap();
+            }
+            self.hash_prevouts = Some(sha256d::Hash::from_engine(enc));
+        }
+        self.hash_prevouts.unwrap()
```

Comment:
I'd personally prefer something like:
```
if let Some(hash) = self.hashprevout {
    hash
} else {
        let mut enc = sha256d::Hash::engine();
    for txin in &self.tx.input {
        txin.previous_output.consensus_encode(&mut enc).unwrap();
    }
    let hash = ha256d::Hash::from_engine(enc));
    self.hash_prevouts = Some(hash);
    hash
}
```
because unwraps have higher maintenance overhead (change to a different line can lead to a panic in corner cases)
but I'm not sure everyone will agree with me.

Reply:
Or even better:
```
    *self.hash_prevouts().get_or_insert_with(|| {
        let mut enc = sha256d::Hash::engine();
        for txin in &self.tx.input {
            txin.previous_output.consensus_encode(&mut enc).unwrap();
        }
        ha256d::Hash::from_engine(enc)
    })
```

Reply:
let me wrestle the borrow checker for a bit and get back to you on the last example

Reply:
ok done